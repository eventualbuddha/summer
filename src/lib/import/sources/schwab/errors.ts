import type { ParseMoneyError } from '$lib/import/parse/money';
import type { Interval } from 'luxon';
import {
	InvalidDateError,
	MissingHeaderError,
	MissingTableCellError,
	ParseStatementSummaryError
} from '../../parse/errors';
import type { StatementSummary } from './summary';
import type { ActivityEntry } from './entries';

export type ImportStatementError =
	| ParseMoneyError
	| ParseActivityEntriesError
	| InvalidTransactionError
	| InvalidStatementIntervalError
	| InvalidStatementSummaryError
	| InconsistentStatementIntervalError
	| ParseStatementSummaryError;

export class InvalidTransactionError extends Error {
	readonly activityEntry: ActivityEntry;

	constructor(message: string, activityEntry: ActivityEntry) {
		super(message);
		this.activityEntry = activityEntry;
	}
}

export class InvalidStatementIntervalError extends Error {
	readonly interval: Interval;

	constructor(interval: Interval) {
		super(`Invalid statement interval: ${interval.invalidExplanation}`);
		this.interval = interval;
	}
}

export class InconsistentStatementIntervalError extends Error {
	readonly firstInterval: Interval;
	readonly secondInterval: Interval;

	constructor(firstInterval: Interval, secondInterval: Interval) {
		super(`Inconsistent statement intervals: ${firstInterval} and ${secondInterval}`);
		this.firstInterval = firstInterval;
		this.secondInterval = secondInterval;
	}
}

export class InvalidStatementSummaryError extends Error {
	readonly summary: StatementSummary;

	readonly summaryValue: number;
	readonly computedValue: number;

	constructor(
		message: string,
		summary: StatementSummary,
		summaryValue: number,
		computedValue: number
	) {
		super(message);
		this.summary = summary;
		this.summaryValue = summaryValue;
		this.computedValue = computedValue;
	}
}

export class InvalidMoneyAmountError extends Error {
	readonly column: string;
	readonly amount: string;
	readonly parseError: ParseMoneyError;

	constructor(column: string, amount: string, parseError: ParseMoneyError) {
		super(`Invalid money amount in column '${column}': ${amount} (${parseError.message})`);
		this.column = column;
		this.amount = amount;
		this.parseError = parseError;
	}
}

export type ParseActivityEntriesError =
	| MissingHeaderError
	| MissingTableCellError
	| InvalidDateError
	| InvalidMoneyAmountError;
