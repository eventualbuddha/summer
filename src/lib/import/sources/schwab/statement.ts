import type { ImportedTransaction } from '$lib/import/ImportedTransaction';
import type { Statement } from '$lib/import/statement/Statement';
import { Result } from '@badrap/result';
import { ParseStatementError } from '../../parse/errors';
import { parseAmount } from '../../parse/money';
import { parseActivityEntries, type ActivityEntry } from './entries';
import { parseStatementPeriodLines, StatementSummary } from './summary';

/**
 * Parses a statement from text extracted from a Schwab statement PDF. Expects a
 * fairly specific format, with transaction data in a table, usually across
 * multiple pages. Also expects a summary section that contains the total
 * debits and credits as well as the ending balance.
 */
export function* parseStatement(
	statement: Statement
): Generator<Result<ImportedTransaction | StatementSummary, ParseStatementError>> {
	// Find the start of the statement summary by looking for "Account Number: XXXX".
	const accountNumberLabel = statement.navigator.find(/Account Number: (.+)/);
	if (!accountNumberLabel) {
		yield Result.err(ParseStatementError.MissingLabel('Account Number'));
		return;
	}

	const accountNumber = accountNumberLabel.text.str.trim().replace(/Account Number: (.+)/, '$1');

	const statementPeriodLabel = accountNumberLabel.pageLocation.findBefore('Statement Period');

	if (!statementPeriodLabel) {
		yield Result.err(
			ParseStatementError.MissingLabel('Statement Period', {
				pageNumber: accountNumberLabel.pageNumber,
				searchFromText: accountNumberLabel.text
			})
		);
		return;
	}

	const firstStatementPeriodLine = statementPeriodLabel.findDown(/\d+/, {
		alignment: 'left',
		maxGap: 10
	});

	if (!firstStatementPeriodLine) {
		yield Result.err(
			ParseStatementError.MissingLabel('Statement Period', {
				pageNumber: statementPeriodLabel.pageNumber,
				searchFromText: statementPeriodLabel.text,
				searchDirection: 'down'
			})
		);
		return;
	}

	const secondStatementPeriodLine = firstStatementPeriodLine.findDown(/\d+/, {
		alignment: 'left',
		maxGap: 10
	});

	const statementPeriod = parseStatementPeriodLines(
		firstStatementPeriodLine.text.str,
		secondStatementPeriodLine?.text.str
	);

	if (statementPeriod.isErr) {
		yield Result.err(statementPeriod.error);
		return;
	}

	const accountName = accountNumberLabel.findLeft(/\w{5,}/);

	if (!accountName) {
		yield Result.err(
			ParseStatementError.MissingValue('Account Name', {
				pageNumber: accountNumberLabel.pageNumber,
				searchFromText: accountNumberLabel.text,
				searchDirection: 'left'
			})
		);
		return;
	}

	const beginningBalance = accountName.findDown('Beginning Balance', { alignment: 'left' });

	if (!beginningBalance) {
		yield Result.err(
			ParseStatementError.MissingLabel('Beginning Balance', {
				pageNumber: accountName.pageNumber,
				searchFromText: accountName.text,
				searchDirection: 'down'
			})
		);
		return;
	}

	const summaryTable = beginningBalance.pageLocation.findTable(
		[
			'Beginning Balance',
			'Deposits and Credits',
			'Interest Paid',
			'Withdrawals and Other Debits',
			'Other Fees',
			'Ending Balance'
		],
		[/\d+/, /\d+/, /\d+/, /\d+/, /\d+/, /\d+/],
		{ labelAlignment: 'left', valueAlignment: 'right', maxGap: 20 }
	);

	if (!summaryTable) {
		yield Result.err(
			ParseStatementError.MissingLabel('Summary', {
				pageNumber: beginningBalance.pageNumber,
				searchFromText: beginningBalance.text
			})
		);
		return;
	}

	const amounts = Result.all([
		parseAmount(summaryTable[0][1].text.str),
		parseAmount(summaryTable[1][1].text.str),
		parseAmount(summaryTable[2][1].text.str),
		parseAmount(summaryTable[3][1].text.str),
		parseAmount(summaryTable[4][1].text.str),
		parseAmount(summaryTable[5][1].text.str)
	]);

	if (amounts.isErr) {
		yield Result.err(
			ParseStatementError.InvalidValue('Invalid summary money amount', {
				pageNumber: beginningBalance.pageNumber,
				cause: amounts.error
			})
		);
		return;
	}

	const summary = new StatementSummary(
		statementPeriod.value,
		accountNumber,
		accountName.text.str.trim(),
		amounts.value[0],
		amounts.value[1],
		amounts.value[2],
		amounts.value[3],
		amounts.value[4],
		amounts.value[5]
	);
	yield Result.ok(summary);

	const parseResults = statement.pages.map((page) =>
		parseActivityEntries(page, statementPeriod.value)
	);
	const entries: ActivityEntry[] = [];

	for (const [index, parseResult] of parseResults.entries()) {
		if (parseResult.isErr) {
			if (entries.length > 0) {
				const allRemainingResultsAreErrors = parseResults
					.slice(index + 1)
					.every(({ isErr }) => isErr);

				if (!allRemainingResultsAreErrors) {
					yield Result.err(
						ParseStatementError.InvalidValue(
							`Expected activity entries to appear on consecutive pages, got entries on pages ${parseResults
								.entries()
								.filter(([, { isOk }]) => isOk)
								.map(([index]) => statement.pages[index]?.pageNumber)
								.toArray()
								.join(', ')}`
						)
					);

					return;
				}
			}
		} else {
			// Include any entries that affect the account balance,
			// i.e. not "Beginning Balance" or "Ending Balance".
			entries.push(...parseResult.value.filter(({ amount }) => !Number.isNaN(amount)));
		}
	}

	const { credits, debits } = entries.reduce(
		(acc, entry) => ({
			credits: acc.credits + (entry.credit ?? 0),
			debits: acc.debits - (entry.debit ?? 0)
		}),
		{ credits: 0, debits: 0 }
	);

	const computedCredits = summary.depositsAndCredits + summary.interestPaid;
	const computedDebits = summary.withdrawalsAndDebits + summary.otherFees;

	if (credits !== computedCredits) {
		yield Result.err(
			ParseStatementError.InvalidValue(
				`Credits do not match the computed value: ${computedCredits} ≠ ${credits}`
			)
		);
		return;
	}

	if (debits !== computedDebits) {
		yield Result.err(
			ParseStatementError.InvalidValue(
				`Debits do not match the computed value: ${computedDebits} ≠ ${debits}`
			)
		);
		return;
	}

	const computedEndingBalance = summary.beginningBalance + credits + debits;

	if (summary.endingBalance !== computedEndingBalance) {
		yield Result.err(
			ParseStatementError.InvalidValue(
				`Ending balance does not match the computed value: ${summary.endingBalance} ≠ ${computedEndingBalance}`
			)
		);
		return;
	}

	for (const entry of entries) {
		yield Result.ok(entry);
	}
}
