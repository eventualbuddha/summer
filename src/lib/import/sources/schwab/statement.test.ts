import { Result } from '@badrap/result';
import { expect, test } from 'vitest';
import * as fixtures from '../../../../../tests/fixtures';
import { InvalidStatementSummaryError } from './errors';
import { parseStatement } from './statement';

test('transactions', async () => {
	const importedTransactions = parseStatement(fixtures.schwab.checking.statement)
		.filter((result) => result.isOk)
		.map((result) => result.value)
		.toArray();

	expect(importedTransactions).toMatchSnapshot();
});

test('summary mismatch', async () => {
	const statementSummaryErrors = parseStatement(fixtures.schwab.checking.statement)
		.filter((result) => result.isErr && result.error instanceof InvalidStatementSummaryError)
		.toArray();
	expect(statementSummaryErrors).toEqual([]);

	for (const page of fixtures.schwab.checking.statement.pages) {
		for (const text of page.texts) {
			if (text.str === '$1,594.08') {
				Object.defineProperty(text, 'str', { value: '$1,594.07', configurable: true });
			}
		}
	}

	const afterModificationErrors = parseStatement(fixtures.schwab.checking.statement)
		.filter((result) => result.isErr && result.error instanceof InvalidStatementSummaryError)
		.toArray();
	expect(afterModificationErrors).toEqual([
		Result.err(
			new InvalidStatementSummaryError(
				'Ending balance does not match the computed value',
				expect.any(Object),
				159407,
				159408
			)
		)
	]);
});
