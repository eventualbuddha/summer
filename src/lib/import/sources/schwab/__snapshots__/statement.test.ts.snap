// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`transactions 1`] = `
[
  StatementSummary {
    "account": "829475629059",
    "accountName": "High Yield Investor Checking",
    "accountType": "banking",
    "beginningBalance": 378295,
    "closingDate": DateTime {
      "_zone": SystemZone {},
      "c": {
        "day": 30,
        "hour": 0,
        "millisecond": 0,
        "minute": 0,
        "month": 11,
        "second": 0,
        "year": 2020,
      },
      "invalid": null,
      "isLuxonDateTime": true,
      "loc": Locale {
        "eraCache": {},
        "fastNumbersCached": null,
        "intl": "en-US",
        "locale": "en-US",
        "meridiemCache": null,
        "monthsCache": {
          "format": {},
          "standalone": {},
        },
        "numberingSystem": null,
        "outputCalendar": null,
        "specifiedLocale": null,
        "weekSettings": null,
        "weekdaysCache": {
          "format": {},
          "standalone": {},
        },
      },
      "localWeekData": null,
      "o": -0,
      "ts": 1606694400000,
      "weekData": null,
    },
    "depositsAndCredits": 180424,
    "endingBalance": 159408,
    "interestPaid": 37,
    "otherFees": 0,
    "period": Interval {
      "e": DateTime {
        "_zone": SystemZone {},
        "c": {
          "day": 30,
          "hour": 0,
          "millisecond": 0,
          "minute": 0,
          "month": 11,
          "second": 0,
          "year": 2020,
        },
        "invalid": null,
        "isLuxonDateTime": true,
        "loc": Locale {
          "eraCache": {},
          "fastNumbersCached": null,
          "intl": "en-US",
          "locale": "en-US",
          "meridiemCache": null,
          "monthsCache": {
            "format": {},
            "standalone": {},
          },
          "numberingSystem": null,
          "outputCalendar": null,
          "specifiedLocale": null,
          "weekSettings": null,
          "weekdaysCache": {
            "format": {},
            "standalone": {},
          },
        },
        "localWeekData": null,
        "o": -0,
        "ts": 1606694400000,
        "weekData": null,
      },
      "invalid": null,
      "isLuxonInterval": true,
      "s": DateTime {
        "_zone": SystemZone {},
        "c": {
          "day": 31,
          "hour": 0,
          "millisecond": 0,
          "minute": 0,
          "month": 10,
          "second": 0,
          "year": 2020,
        },
        "invalid": null,
        "isLuxonDateTime": true,
        "loc": Locale {
          "eraCache": {},
          "fastNumbersCached": null,
          "intl": "en-US",
          "locale": "en-US",
          "meridiemCache": null,
          "monthsCache": {
            "format": {},
            "standalone": {},
          },
          "numberingSystem": null,
          "outputCalendar": null,
          "specifiedLocale": null,
          "weekSettings": null,
          "weekdaysCache": {
            "format": {},
            "standalone": {},
          },
        },
        "localWeekData": null,
        "o": -0,
        "ts": 1604102400000,
        "weekData": null,
      },
    },
    "withdrawalsAndDebits": -399348,
  },
  ActivityEntry {
    "amount": -28450,
    "balance": 349845,
    "credit": undefined,
    "date": DateTime {
      "_zone": SystemZone {},
      "c": {
        "day": 2,
        "hour": 0,
        "millisecond": 0,
        "minute": 0,
        "month": 11,
        "second": 0,
        "year": 2020,
      },
      "invalid": null,
      "isLuxonDateTime": true,
      "loc": Locale {
        "eraCache": {},
        "fastNumbersCached": null,
        "intl": "en-US",
        "locale": "en-US",
        "meridiemCache": null,
        "monthsCache": {
          "format": {},
          "standalone": {},
        },
        "numberingSystem": null,
        "outputCalendar": null,
        "specifiedLocale": null,
        "weekSettings": null,
        "weekdaysCache": {
          "format": {},
          "standalone": {},
        },
      },
      "localWeekData": null,
      "o": -0,
      "ts": 1604275200000,
      "weekData": null,
    },
    "debit": 28450,
    "description": "MUGGLECARD GGANK PAYMENT 103120",
    "kind": "Charge",
    "pageNumber": 3,
    "statementDescription": "Electronic Withdrawal MUGGLECARD GGANK PAYMENT 103120",
    "type": "Electronic Withdrawal",
  },
  ActivityEntry {
    "amount": -7500,
    "balance": 342345,
    "credit": undefined,
    "date": DateTime {
      "_zone": SystemZone {},
      "c": {
        "day": 2,
        "hour": 0,
        "millisecond": 0,
        "minute": 0,
        "month": 11,
        "second": 0,
        "year": 2020,
      },
      "invalid": null,
      "isLuxonDateTime": true,
      "loc": Locale {
        "eraCache": {},
        "fastNumbersCached": null,
        "intl": "en-US",
        "locale": "en-US",
        "meridiemCache": null,
        "monthsCache": {
          "format": {},
          "standalone": {},
        },
        "numberingSystem": null,
        "outputCalendar": null,
        "specifiedLocale": null,
        "weekSettings": null,
        "weekdaysCache": {
          "format": {},
          "standalone": {},
        },
      },
      "localWeekData": null,
      "o": -0,
      "ts": 1604275200000,
      "weekData": null,
    },
    "debit": 7500,
    "description": "Check Paid #995070",
    "kind": "Check",
    "pageNumber": 3,
    "statementDescription": "Check Check Paid #995070",
    "type": "Check",
  },
  ActivityEntry {
    "amount": -10822,
    "balance": 331523,
    "credit": undefined,
    "date": DateTime {
      "_zone": SystemZone {},
      "c": {
        "day": 2,
        "hour": 0,
        "millisecond": 0,
        "minute": 0,
        "month": 11,
        "second": 0,
        "year": 2020,
      },
      "invalid": null,
      "isLuxonDateTime": true,
      "loc": Locale {
        "eraCache": {},
        "fastNumbersCached": null,
        "intl": "en-US",
        "locale": "en-US",
        "meridiemCache": null,
        "monthsCache": {
          "format": {},
          "standalone": {},
        },
        "numberingSystem": null,
        "outputCalendar": null,
        "specifiedLocale": null,
        "weekSettings": null,
        "weekdaysCache": {
          "format": {},
          "standalone": {},
        },
      },
      "localWeekData": null,
      "o": -0,
      "ts": 1604275200000,
      "weekData": null,
    },
    "debit": 10822,
    "description": "CASH APP*RON WEASL8472364892, CA #000000822048472",
    "kind": "Charge",
    "pageNumber": 3,
    "statementDescription": "Visa Debit Card Purchase CASH APP*RON WEASL8472364892, CA #000000822048472",
    "type": "Visa Debit Card Purchase",
  },
  ActivityEntry {
    "amount": 150000,
    "balance": 481523,
    "credit": 150000,
    "date": DateTime {
      "_zone": SystemZone {},
      "c": {
        "day": 2,
        "hour": 0,
        "millisecond": 0,
        "minute": 0,
        "month": 11,
        "second": 0,
        "year": 2020,
      },
      "invalid": null,
      "isLuxonDateTime": true,
      "loc": Locale {
        "eraCache": {},
        "fastNumbersCached": null,
        "intl": "en-US",
        "locale": "en-US",
        "meridiemCache": null,
        "monthsCache": {
          "format": {},
          "standalone": {},
        },
        "numberingSystem": null,
        "outputCalendar": null,
        "specifiedLocale": null,
        "weekSettings": null,
        "weekdaysCache": {
          "format": {},
          "standalone": {},
        },
      },
      "localWeekData": null,
      "o": -0,
      "ts": 1604275200000,
      "weekData": null,
    },
    "debit": undefined,
    "description": "Transfer from High Yield Checking -5125",
    "kind": "Transfer",
    "pageNumber": 3,
    "statementDescription": "Money Transfer Transfer from High Yield Checking -5125",
    "type": "Money Transfer",
  },
  ActivityEntry {
    "amount": -111111,
    "balance": 370412,
    "credit": undefined,
    "date": DateTime {
      "_zone": SystemZone {},
      "c": {
        "day": 3,
        "hour": 0,
        "millisecond": 0,
        "minute": 0,
        "month": 11,
        "second": 0,
        "year": 2020,
      },
      "invalid": null,
      "isLuxonDateTime": true,
      "loc": Locale {
        "eraCache": {},
        "fastNumbersCached": null,
        "intl": "en-US",
        "locale": "en-US",
        "meridiemCache": null,
        "monthsCache": {
          "format": {},
          "standalone": {},
        },
        "numberingSystem": null,
        "outputCalendar": null,
        "specifiedLocale": null,
        "weekSettings": null,
        "weekdaysCache": {
          "format": {},
          "standalone": {},
        },
      },
      "localWeekData": null,
      "o": -0,
      "ts": 1604361600000,
      "weekData": null,
    },
    "debit": 111111,
    "description": "QUICKEN LOANS MTG PYMTS 832047",
    "kind": "Charge",
    "pageNumber": 4,
    "statementDescription": "Electronic Withdrawal QUICKEN LOANS MTG PYMTS 832047",
    "type": "Electronic Withdrawal",
  },
  ActivityEntry {
    "amount": -8000,
    "balance": 362412,
    "credit": undefined,
    "date": DateTime {
      "_zone": SystemZone {},
      "c": {
        "day": 9,
        "hour": 0,
        "millisecond": 0,
        "minute": 0,
        "month": 11,
        "second": 0,
        "year": 2020,
      },
      "invalid": null,
      "isLuxonDateTime": true,
      "loc": Locale {
        "eraCache": {},
        "fastNumbersCached": null,
        "intl": "en-US",
        "locale": "en-US",
        "meridiemCache": null,
        "monthsCache": {
          "format": {},
          "standalone": {},
        },
        "numberingSystem": null,
        "outputCalendar": null,
        "specifiedLocale": null,
        "weekSettings": null,
        "weekdaysCache": {
          "format": {},
          "standalone": {},
        },
      },
      "localWeekData": null,
      "o": -0,
      "ts": 1604880000000,
      "weekData": null,
    },
    "debit": 8000,
    "description": "Check Paid #995071",
    "kind": "Check",
    "pageNumber": 4,
    "statementDescription": "Check Check Paid #995071",
    "type": "Check",
  },
  ActivityEntry {
    "amount": 30424,
    "balance": 392836,
    "credit": 30424,
    "date": DateTime {
      "_zone": SystemZone {},
      "c": {
        "day": 13,
        "hour": 0,
        "millisecond": 0,
        "minute": 0,
        "month": 11,
        "second": 0,
        "year": 2020,
      },
      "invalid": null,
      "isLuxonDateTime": true,
      "loc": Locale {
        "eraCache": {},
        "fastNumbersCached": null,
        "intl": "en-US",
        "locale": "en-US",
        "meridiemCache": null,
        "monthsCache": {
          "format": {},
          "standalone": {},
        },
        "numberingSystem": null,
        "outputCalendar": null,
        "specifiedLocale": null,
        "weekSettings": null,
        "weekdaysCache": {
          "format": {},
          "standalone": {},
        },
      },
      "localWeekData": null,
      "o": -0,
      "ts": 1605225600000,
      "weekData": null,
    },
    "debit": undefined,
    "description": "Square Inc * Cash App 949257",
    "kind": "Deposit",
    "pageNumber": 4,
    "statementDescription": "Electronic Deposit Square Inc * Cash App 949257",
    "type": "Electronic Deposit",
  },
  ActivityEntry {
    "amount": -49401,
    "balance": 343435,
    "credit": undefined,
    "date": DateTime {
      "_zone": SystemZone {},
      "c": {
        "day": 17,
        "hour": 0,
        "millisecond": 0,
        "minute": 0,
        "month": 11,
        "second": 0,
        "year": 2020,
      },
      "invalid": null,
      "isLuxonDateTime": true,
      "loc": Locale {
        "eraCache": {},
        "fastNumbersCached": null,
        "intl": "en-US",
        "locale": "en-US",
        "meridiemCache": null,
        "monthsCache": {
          "format": {},
          "standalone": {},
        },
        "numberingSystem": null,
        "outputCalendar": null,
        "specifiedLocale": null,
        "weekSettings": null,
        "weekdaysCache": {
          "format": {},
          "standalone": {},
        },
      },
      "localWeekData": null,
      "o": -0,
      "ts": 1605571200000,
      "weekData": null,
    },
    "debit": 49401,
    "description": "AMEX EPAYMENT ACH PMT 842780",
    "kind": "Charge",
    "pageNumber": 4,
    "statementDescription": "Electronic Withdrawal AMEX EPAYMENT ACH PMT 842780",
    "type": "Electronic Withdrawal",
  },
  ActivityEntry {
    "amount": -2011,
    "balance": 341424,
    "credit": undefined,
    "date": DateTime {
      "_zone": SystemZone {},
      "c": {
        "day": 18,
        "hour": 0,
        "millisecond": 0,
        "minute": 0,
        "month": 11,
        "second": 0,
        "year": 2020,
      },
      "invalid": null,
      "isLuxonDateTime": true,
      "loc": Locale {
        "eraCache": {},
        "fastNumbersCached": null,
        "intl": "en-US",
        "locale": "en-US",
        "meridiemCache": null,
        "monthsCache": {
          "format": {},
          "standalone": {},
        },
        "numberingSystem": null,
        "outputCalendar": null,
        "specifiedLocale": null,
        "weekSettings": null,
        "weekdaysCache": {
          "format": {},
          "standalone": {},
        },
      },
      "localWeekData": null,
      "o": -0,
      "ts": 1605657600000,
      "weekData": null,
    },
    "debit": 2011,
    "description": "FLOURISH AND BLOTTS 203",
    "kind": "Charge",
    "pageNumber": 4,
    "statementDescription": "Electronic Withdrawal FLOURISH AND BLOTTS 203",
    "type": "Electronic Withdrawal",
  },
  ActivityEntry {
    "amount": -1200,
    "balance": 340224,
    "credit": undefined,
    "date": DateTime {
      "_zone": SystemZone {},
      "c": {
        "day": 18,
        "hour": 0,
        "millisecond": 0,
        "minute": 0,
        "month": 11,
        "second": 0,
        "year": 2020,
      },
      "invalid": null,
      "isLuxonDateTime": true,
      "loc": Locale {
        "eraCache": {},
        "fastNumbersCached": null,
        "intl": "en-US",
        "locale": "en-US",
        "meridiemCache": null,
        "monthsCache": {
          "format": {},
          "standalone": {},
        },
        "numberingSystem": null,
        "outputCalendar": null,
        "specifiedLocale": null,
        "weekSettings": null,
        "weekdaysCache": {
          "format": {},
          "standalone": {},
        },
      },
      "localWeekData": null,
      "o": -0,
      "ts": 1605657600000,
      "weekData": null,
    },
    "debit": 1200,
    "description": "Check Paid #1202",
    "kind": "Check",
    "pageNumber": 4,
    "statementDescription": "Check Check Paid #1202",
    "type": "Check",
  },
  ActivityEntry {
    "amount": -23355,
    "balance": 316869,
    "credit": undefined,
    "date": DateTime {
      "_zone": SystemZone {},
      "c": {
        "day": 19,
        "hour": 0,
        "millisecond": 0,
        "minute": 0,
        "month": 11,
        "second": 0,
        "year": 2020,
      },
      "invalid": null,
      "isLuxonDateTime": true,
      "loc": Locale {
        "eraCache": {},
        "fastNumbersCached": null,
        "intl": "en-US",
        "locale": "en-US",
        "meridiemCache": null,
        "monthsCache": {
          "format": {},
          "standalone": {},
        },
        "numberingSystem": null,
        "outputCalendar": null,
        "specifiedLocale": null,
        "weekSettings": null,
        "weekdaysCache": {
          "format": {},
          "standalone": {},
        },
      },
      "localWeekData": null,
      "o": -0,
      "ts": 1605744000000,
      "weekData": null,
    },
    "debit": 23355,
    "description": "LEAKY CAULDRON BB 84397",
    "kind": "Charge",
    "pageNumber": 4,
    "statementDescription": "Electronic Withdrawal LEAKY CAULDRON BB 84397",
    "type": "Electronic Withdrawal",
  },
  ActivityEntry {
    "amount": -10000,
    "balance": 326869,
    "credit": undefined,
    "date": DateTime {
      "_zone": SystemZone {},
      "c": {
        "day": 21,
        "hour": 0,
        "millisecond": 0,
        "minute": 0,
        "month": 11,
        "second": 0,
        "year": 2020,
      },
      "invalid": null,
      "isLuxonDateTime": true,
      "loc": Locale {
        "eraCache": {},
        "fastNumbersCached": null,
        "intl": "en-US",
        "locale": "en-US",
        "meridiemCache": null,
        "monthsCache": {
          "format": {},
          "standalone": {},
        },
        "numberingSystem": null,
        "outputCalendar": null,
        "specifiedLocale": null,
        "weekSettings": null,
        "weekdaysCache": {
          "format": {},
          "standalone": {},
        },
      },
      "localWeekData": null,
      "o": -0,
      "ts": 1605916800000,
      "weekData": null,
    },
    "debit": 10000,
    "description": undefined,
    "kind": "Deposit",
    "pageNumber": 4,
    "statementDescription": "Deposit Mobile Banking",
    "type": "Deposit Mobile Banking",
  },
  ActivityEntry {
    "amount": -9000,
    "balance": 317869,
    "credit": undefined,
    "date": DateTime {
      "_zone": SystemZone {},
      "c": {
        "day": 23,
        "hour": 0,
        "millisecond": 0,
        "minute": 0,
        "month": 11,
        "second": 0,
        "year": 2020,
      },
      "invalid": null,
      "isLuxonDateTime": true,
      "loc": Locale {
        "eraCache": {},
        "fastNumbersCached": null,
        "intl": "en-US",
        "locale": "en-US",
        "meridiemCache": null,
        "monthsCache": {
          "format": {},
          "standalone": {},
        },
        "numberingSystem": null,
        "outputCalendar": null,
        "specifiedLocale": null,
        "weekSettings": null,
        "weekdaysCache": {
          "format": {},
          "standalone": {},
        },
      },
      "localWeekData": null,
      "o": -0,
      "ts": 1606089600000,
      "weekData": null,
    },
    "debit": 9000,
    "description": "VENMO PAYMENT",
    "kind": "Charge",
    "pageNumber": 4,
    "statementDescription": "Electronic Withdrawal VENMO PAYMENT",
    "type": "Electronic Withdrawal",
  },
  ActivityEntry {
    "amount": -15000,
    "balance": 302869,
    "credit": undefined,
    "date": DateTime {
      "_zone": SystemZone {},
      "c": {
        "day": 23,
        "hour": 0,
        "millisecond": 0,
        "minute": 0,
        "month": 11,
        "second": 0,
        "year": 2020,
      },
      "invalid": null,
      "isLuxonDateTime": true,
      "loc": Locale {
        "eraCache": {},
        "fastNumbersCached": null,
        "intl": "en-US",
        "locale": "en-US",
        "meridiemCache": null,
        "monthsCache": {
          "format": {},
          "standalone": {},
        },
        "numberingSystem": null,
        "outputCalendar": null,
        "specifiedLocale": null,
        "weekSettings": null,
        "weekdaysCache": {
          "format": {},
          "standalone": {},
        },
      },
      "localWeekData": null,
      "o": -0,
      "ts": 1606089600000,
      "weekData": null,
    },
    "debit": 15000,
    "description": "VENMO PAYMENT",
    "kind": "Charge",
    "pageNumber": 4,
    "statementDescription": "Electronic Withdrawal VENMO PAYMENT",
    "type": "Electronic Withdrawal",
  },
  ActivityEntry {
    "amount": -100000,
    "balance": 402869,
    "credit": undefined,
    "date": DateTime {
      "_zone": SystemZone {},
      "c": {
        "day": 25,
        "hour": 0,
        "millisecond": 0,
        "minute": 0,
        "month": 11,
        "second": 0,
        "year": 2020,
      },
      "invalid": null,
      "isLuxonDateTime": true,
      "loc": Locale {
        "eraCache": {},
        "fastNumbersCached": null,
        "intl": "en-US",
        "locale": "en-US",
        "meridiemCache": null,
        "monthsCache": {
          "format": {},
          "standalone": {},
        },
        "numberingSystem": null,
        "outputCalendar": null,
        "specifiedLocale": null,
        "weekSettings": null,
        "weekdaysCache": {
          "format": {},
          "standalone": {},
        },
      },
      "localWeekData": null,
      "o": -0,
      "ts": 1606262400000,
      "weekData": null,
    },
    "debit": 100000,
    "description": undefined,
    "kind": "Deposit",
    "pageNumber": 4,
    "statementDescription": "Deposit Mobile Banking",
    "type": "Deposit Mobile Banking",
  },
  ActivityEntry {
    "amount": -100,
    "balance": 402769,
    "credit": undefined,
    "date": DateTime {
      "_zone": SystemZone {},
      "c": {
        "day": 27,
        "hour": 0,
        "millisecond": 0,
        "minute": 0,
        "month": 11,
        "second": 0,
        "year": 2020,
      },
      "invalid": null,
      "isLuxonDateTime": true,
      "loc": Locale {
        "eraCache": {},
        "fastNumbersCached": null,
        "intl": "en-US",
        "locale": "en-US",
        "meridiemCache": null,
        "monthsCache": {
          "format": {},
          "standalone": {},
        },
        "numberingSystem": null,
        "outputCalendar": null,
        "specifiedLocale": null,
        "weekSettings": null,
        "weekdaysCache": {
          "format": {},
          "standalone": {},
        },
      },
      "localWeekData": null,
      "o": -0,
      "ts": 1606435200000,
      "weekData": null,
    },
    "debit": 100,
    "description": "CITY OF LONDON PARKING GARAGE #000000190070441",
    "kind": "Charge",
    "pageNumber": 4,
    "statementDescription": "Visa Debit Card Purchase CITY OF LONDON PARKING GARAGE #000000190070441",
    "type": "Visa Debit Card Purchase",
  },
  ActivityEntry {
    "amount": -23398,
    "balance": 379371,
    "credit": undefined,
    "date": DateTime {
      "_zone": SystemZone {},
      "c": {
        "day": 30,
        "hour": 0,
        "millisecond": 0,
        "minute": 0,
        "month": 11,
        "second": 0,
        "year": 2020,
      },
      "invalid": null,
      "isLuxonDateTime": true,
      "loc": Locale {
        "eraCache": {},
        "fastNumbersCached": null,
        "intl": "en-US",
        "locale": "en-US",
        "meridiemCache": null,
        "monthsCache": {
          "format": {},
          "standalone": {},
        },
        "numberingSystem": null,
        "outputCalendar": null,
        "specifiedLocale": null,
        "weekSettings": null,
        "weekdaysCache": {
          "format": {},
          "standalone": {},
        },
      },
      "localWeekData": null,
      "o": -0,
      "ts": 1606694400000,
      "weekData": null,
    },
    "debit": 23398,
    "description": "HOGWARTS SCHOOL WWPAY 201127",
    "kind": "Charge",
    "pageNumber": 5,
    "statementDescription": "Electronic Withdrawal HOGWARTS SCHOOL WWPAY 201127",
    "type": "Electronic Withdrawal",
  },
  ActivityEntry {
    "amount": 37,
    "balance": 379408,
    "credit": 37,
    "date": DateTime {
      "_zone": SystemZone {},
      "c": {
        "day": 30,
        "hour": 0,
        "millisecond": 0,
        "minute": 0,
        "month": 11,
        "second": 0,
        "year": 2020,
      },
      "invalid": null,
      "isLuxonDateTime": true,
      "loc": Locale {
        "eraCache": {},
        "fastNumbersCached": null,
        "intl": "en-US",
        "locale": "en-US",
        "meridiemCache": null,
        "monthsCache": {
          "format": {},
          "standalone": {},
        },
        "numberingSystem": null,
        "outputCalendar": null,
        "specifiedLocale": null,
        "weekSettings": null,
        "weekdaysCache": {
          "format": {},
          "standalone": {},
        },
      },
      "localWeekData": null,
      "o": -0,
      "ts": 1606694400000,
      "weekData": null,
    },
    "debit": undefined,
    "description": undefined,
    "kind": "Interest",
    "pageNumber": 5,
    "statementDescription": "Interest Paid",
    "type": "Interest Paid",
  },
]
`;
